### PR TITLE
Max entropy

### LINKED ISSUE #3558: Add `max_entropy` function
### Feature details

It would be nice to add support for $H_{max}(ρ)=\log(rank(ρ))$ [max entropy](http://www.scholarpedia.org/article/Quantum_entropies#Max_entropy_.28Hartley_entropy.29)  in the quantum information module. This can be achieved by implementing a separate function in `pennylane/math/quantum.py` that takes as input either a state vector or a density matrix, plus additional arguments such as the logarithm base, and outputs a scalar.

Of course, we would want the function to be fully differentiable as well.

### Implementation

**Steps to take:**
1. Implement a function `max_entropy()` in `pennylane/math/quantum.py` that takes as input a state vector or a density matrix, plus any additional arguments you see fit.
2. Add tests for the function in `tests/math/test_entropies_math.py`.

**Notes:**
1. For both of these steps, you can refer to the implementations of the other quantum information functions.
2. Other quantum information quantities have additional support for i) measurement processes and ii) QNode transforms. In this case neither of these are required.
3. Differentiability is required. Testing must cover differentiability for all interfaces (autograd, TF, torch, and JAX). (Note that tests for differentiability for the other entropies, such as `vn_entropy`, are instead done in `tests/qinfo/test_entropies.py`. Since you're not implementing a QNode transform, it's fine to place these tests in `tests/math/test_entropies_math.py` instead.)

### How important would you say this feature is?

1: Not important. Would be nice to have.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by BorjaRequena**
**Context:**

The goal is to add a `max_entropy()` function to `pennylane/math/quantum.py`. This function computes the maximum entropy of an given state (state vector or matrix) defined as $H_{\text{max}}=\log\left(\text{rank}(\rho)\right)$.

**Description of the Change:**

I have defined the function following the same style as the other entropy-related ones in `pennylane/math/quantum.py`. It accepts an input state in the form of a state vector or a density matrix. As additional arguments, it takes the same ones as the other functions in the same module: a list of indices for the subsystems to consider, the logarithm basis (defaults to natural), a Boolean option to check the state is valid (defaults to false), and the data type of the tensors (defaults to complex 128).

To perform the calculation, I first build the reduced density matrix of the subsystem to consider with the `reduced_dm()` function from the same module. Then, I compute the rank of the resulting matrix with  `qml.math.linalg.matrix_rank()` (should I use numpy's `np.linalg.matrix_rank()` instead?) and take the logarithm.

I have also included tests that, so far, cover the basic usage of this new function in `tests/math/test_entropies_math.py` in the `TestMaxEntropy` class. The tests consider states with rank 1 and 2 density matrices, both state vector and density matrix inputs, different logarithm bases and all possible interfaces. They are very similar to the Von Neumann entropy tests. 

**Benefits:**

We can easily compute the maximum entropy of quantum states.

**To do:**

I have yet to add tests that cover the differentiability of the function for all interfaces. I also need to check the docs and the tests (I currently can't run them locally so I'll rely on the CI if that's ok), and I need to add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the change, and including a link back to this PR.

**Drawbacks:**

The maximum entropy is a discrete quantity, as the rank of the density matrix will change sharply as a function of the parameters, could this result in differentiability issues?

**Related GitHub Issues:**

In this PR I will eventually resolve #3558 (still work in progress)

**Comments:**

I would specially appreciate some guidance regarding the differentiability of the function and its tests, which I guess they will mainly depend on the function's design. The differentiability tests for the other entropy functions, such as the `vn_entropy()` (from `qml.qinfo`), use QNodes that contain a simple parametrized circuit. Should I follow the same procedure or should I simply consider parametrized input tensors? Going down the QNode path would require some additional support for those, which the reference issue specifically mentions are not required.

Also, I am not sure whether the testing that I have already put in place is thorough enough. Please, let me know whether I should try out some additional cases.  All feedback is more than welcome :)

### CODE REVIEW

**Comment by BorjaRequena** [comment]
I saw there were some issues with tensorflow and `qml.math.linalg.matrix_rank()`. I have now changed the rank calculation by 'counting' the non-zero eigenvalues (those >= 1e-8). The counting consists on dividing all eigenvalues by themselves substituting the dividing zeros by ones, and adding them all up. This also allows parameters that require gradient to go through. For example, we can take the gradient of a parametrized state in torch this way:
```python
import pennylane as qml
import torch

n_qubits = 2
torch.manual_seed(7)
parameters = torch.randn(2**n_qubits, requires_grad=True) # 4 parameters
state_torch = parameters / (parameters**2).sum().sqrt()   # Normalized state
max_entropy = qml.math.max_entropy(state_torch, [1])      # qubit 1 max entropy: log(2)
max_entropy.backward()
parameters.grad # tensor([0., 0., 0., 0.])
```
However, the gradient is always zero.

Let me know if this is the desired behaviour :)

File: pennylane/math/quantum.py
Comment (rmoyard) [inline pennylane/math/quantum.py:1111]: Nice it is a clever implementation and it should work with all interfaces!

File: pennylane/math/quantum.py
Comment (rmoyard) [inline pennylane/math/quantum.py:1073]: Nice examples! Could you add a small comparison with a VN entropy value and show that the maximal entropy is always greater?

**Review by rmoyard — COMMENTED** [review]
Hey @BorjaRequena ! Thanks for your contribution, it looks great already! A couple of comments and answers to your question.

From your code that exactly what we expect. The max entropy is always constant (mixed) or 0 (pure), so there is some parameters where it will be Nan.
```
import pennylane as qml
import torch

n_qubits = 2
torch.manual_seed(7)
parameters = torch.randn(2**n_qubits, requires_grad=True) # 4 parameters
state_torch = parameters / (parameters**2).sum().sqrt()   # Normalized state
max_entropy = qml.math.max_entropy(state_torch, [1])      # qubit 1 max entropy: log(2)
max_entropy.backward()
parameters.grad # tensor([0., 0., 0., 0.])
```
But as you said the full integration requires either to add the QNode integration, or a measurement but it not in the scope of this PR. So I would suggest to simply add a test for each interface like you did in this example and check that the derivative is 0. The integration will done in another PR.

File: pennylane/math/quantum.py
Comment (BorjaRequena) [reply pennylane/math/quantum.py:1073]: Thanks! I added ann additional example comparing with VN entropy. Let me know what you think

File: pennylane/math/quantum.py
Comment (BorjaRequena) [reply pennylane/math/quantum.py:1125]: Nice, thanks! :)

**Comment by BorjaRequena** [comment]
I have included differentiability tests for all the interfaces. These cover all possible input options: state vector / density matrix, indices, different bases, etc.

There was an issue with autograd when `check_state=True` due to the diagonalization in `_check_density_matrix`, which called `np.linalg.eigvalsh` to obtain the eigenvalues. I fixed it by calling `qml.math.linalg.eigh` instead.

For the jax testing, I have tried to also jit the gradient function, but I haven't found a consistent way to do it. The main issue is the indices argument, which conditions the size of the resulting density matrix. Given that it is a list, it is unhashable so I can't mark it as a static argument. Should I try to put more effort on testing this or is it ok to just test the 'basic jax'?

**Comment by rmoyard** [comment]
I was able to solve the problem with jitting:

```
gradient = jax.jit(max_entropy_grad, static_argnums=[1, 2, 3])(params, tuple(wires), base, check_state)
```
basically the arguments to be jitted are usually only the parameters therefore you can specify that you only want to trace them and not the wires and other things.

Another point is to add the documentation, you simply have to add the `max_entropy` to the init file of the math module. Let me know if you need some help.

After those two changes I will be happy to approve the PR, great job!

**Comment by BorjaRequena** [comment]
I have implemented both changes. The tests are passing and the docs seem to render nicely. However, I am not sure about the notation for the maximum entropy, it's currently $S(\rho)_0$ but maybe we could put something more intuitive like $S{\max}(\rho)$?

**Comment by rmoyard** [comment]
Thanks @BorjaRequena yes it is a good solution to use max in subscript. I will approve the PR after the change!

**Comment by BorjaRequena** [comment]
Done! :)

**Review by rmoyard — APPROVED** [review]
Great contribution @BorjaRequena :100: Feel free to merge the PR when the tests are finished.

**Comment by BorjaRequena** [comment]
All tests have passed but I don't think I have write access to merge it :( Can you do it for me, @rmoyard?

**Comment by co9olguy** [comment]
Awesome, thanks @BorjaRequena!

**Pull request merged**
